### ISSUE TITLE
[Plz help] unexpected results from Barracuda with pre-trained model

### ISSUE THREAD

**Issue by JetttZHENG** [Unity-Technologies/ml-agents#2027]
Hello, guys, I would like to try pre-trained Keras Model with  Barracuda, something wrong with the output results, could someone help? Many thanks. Details following:
**1. I have a pre-trained Keras model (Mnist model from google colab).** 
h5 model link: [https://www.dropbox.com/s/iztjy85970n8co2/model.h5?dl=0](url)

**2.  converted the h5 model to pb model using these tool (no errors occurring)**  
Tool Link: [https://www.dropbox.com/s/iztjy85970n8co2/model.h5?dl=0](url)
pb model Link: [https://www.dropbox.com/s/q9095bs7bx1iyjo/model.pb?dl=0](url)

**3.  converted the pb model to nn model using tensorflow_to_barracuda. (no errors occurring), changed its name from model.nn to model.bytes.** 
bytes model link: [https://www.dropbox.com/s/gqelh58aocnpmxn/model.bytes?dl=0](url)

**4.  followed Barracuda.md,  I created a script to get the output of the model.bytes model. 
The shape of the output tensor is correct (1,1,1,10), but values are wrong. 
I also checked the input tensor and it is correct (1,28,28,1) (one grey image with 28*28 pixels).**

Script link: [https://www.dropbox.com/s/i983x9fydjtm0yd/TestTensorflow.cs?dl=0](url)

5. From the message of console, I noticed that there might be something wrong with the **input dictionary** or  **my converted model**.

6. Test Image: 
![mnist_test](https://user-images.githubusercontent.com/43242747/57504190-1e8cb480-7326-11e9-9379-55feea8f7bde.png)

**Comment by JetttZHENG** [comment]
`using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using MLAgents;
using Barracuda;
using System.IO;
using UnityEngine.UI;

public class TestTensorflow : MonoBehaviour
{
    private string loadpath = "Assets/Materials/mnist_test.png";
    // Start is called before the first frame update
    void Start()
    {
        Tensor output_nn = InferenceTest(loadpath);
        
    }
    
    private Tensor InferenceTest(string loadpath)
    {
        Texture2D input_texture2D = LoadByIO(loadpath);
        // Color[] picture_color = input_texture2D.GetPixels();

        var inputs = new Dictionary<string, Tensor>();

        // Load Model into Barracuda
        var model = ModelLoader.LoadFromStreamingAssets("model"+ ".bytes");

        // Create inference engine (Worker)
        var worker = BarracudaWorkerFactory.CreateWorker(BarracudaWorkerFactory.Type.ComputeFast, model);

        //Create inputs
        Tensor tensor = new Tensor(input_texture2D,1);
        tensor.Print();
        Debug.Log(tensor[0,12,12,0]);
        inputs.Add("x", tensor);

        //Execute the model
        worker.Execute(inputs);

        //Fetch outputs
        var output_nn = worker.Fetch();





        output_nn.Dispose();
        worker.Dispose();

        print(output_nn[0, 0, 0, 0]);
        print(output_nn[0, 0, 0, 1]);
        print(output_nn[0, 0, 0, 2]);
        print(output_nn[0, 0, 0, 3]);
        print(output_nn[0, 0, 0, 4]);
        print(output_nn[0, 0, 0, 5]);
        print(output_nn[0, 0, 0, 6]);
        print(output_nn[0, 0, 0, 7]);
        print(output_nn[0, 0, 0, 8]);
        print(output_nn[0, 0, 0, 9]);
        output_nn.Print();

        return output_nn;
    }



    private Texture2D LoadByIO(string loadpath)
    {
        //FileStream creat
        FileStream fileStream = new FileStream(loadpath, FileMode.Open, FileAccess.Read);
        fileStream.Seek(0, SeekOrigin.Begin);
        //
        byte[] bytes = new byte[fileStream.Length];
        //FileStream read
        fileStream.Read(bytes, 0, (int)fileStream.Length);
        //FileStream close
        fileStream.Close();
        fileStream.Dispose();
        fileStream = null;

        // Ceate Texture
        int width = 28;
        int height = 28;
        Texture2D texture2D = new Texture2D(width, height);
        var is_load = texture2D.LoadImage(bytes);
        if (!is_load)
        {
            Debug.Log("Texture Fail");
        }
        return texture2D;
    }
    // Update is called once per frame
    void Update()
    {
        
    }
}
`

**Comment by KozykinRoman** [comment]
Hi, JetttZheng.
I just saw your code. And it looks you called the print functions after dispose the array.
ex. print(output_nn[0, 0, 0, 0]);
So I suggest you put before you release the memory.

**Comment by KozykinRoman** [comment]
And on the other hand, I am suffering from can't find the onnx_to_barracuda.py file.
I hope you help me for this issue.
I already posted the issue on the forum.
https://github.com/Unity-Technologies/ml-agents/issues/2026

**Comment by JetttZHENG** [comment]
> Hi, JetttZheng.
> I just saw your code. And it seems you called the print functions after dispose the array.
> ex. print(output_nn[0, 0, 0, 0]);
> So I suggest you put before you release the memory.

I tried it as you said, but  the values are still not correct. Thanks anyway.

And  I  haven't found the onnx_to_barracuda.py file...

**Comment by JetttZHENG** [comment]
![image](https://user-images.githubusercontent.com/43242747/57523505-ba82e400-7357-11e9-9f04-35cc2c397dc9.png)
![image](https://user-images.githubusercontent.com/43242747/57523529-ca9ac380-7357-11e9-85f2-6619bde7e449.png)

**Comment by JetttZHENG** [comment]
`using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using MLAgents;
using Barracuda;
using System.IO;
using UnityEngine.UI;

public class TestTensorflow : MonoBehaviour
{
    private string loadpath = "Assets/Materials/mnist_test.png";
    // Start is called before the first frame update
    void Start()
    {
        Tensor output_nn = InferenceTest(loadpath);
        
    }
    
    private Tensor InferenceTest(string loadpath)
    {
        Texture2D input_texture2D = LoadByIO(loadpath);
        // Color[] picture_color = input_texture2D.GetPixels();

        var inputs = new Dictionary<string, Tensor>();

        // Load Model into Barracuda
        var model = ModelLoader.LoadFromStreamingAssets("model" + ".bytes");

        // Create inference engine (Worker)
        var worker = BarracudaWorkerFactory.CreateWorker(BarracudaWorkerFactory.Type.ComputeFast, model);

        //Create inputs
        Tensor tensor = new Tensor(input_texture2D,1);
        tensor.Print();
        Debug.Log(tensor[0,12,12,0]);
        inputs.Add("x", tensor);

        //Execute the model
        worker.Execute(inputs);

        //Fetch outputs
        var output_nn = worker.Fetch();

        output_nn.Print();
        Debug.Log(output_nn[0, 0, 0, 0]);
        Debug.Log(output_nn[0, 0, 0, 1]);
        Debug.Log(output_nn[0, 0, 0, 2]);
        Debug.Log(output_nn[0, 0, 0, 3]);
        Debug.Log(output_nn[0, 0, 0, 4]);
        Debug.Log(output_nn[0, 0, 0, 5]);
        Debug.Log(output_nn[0, 0, 0, 6]);
        Debug.Log(output_nn[0, 0, 0, 7]);
        Debug.Log(output_nn[0, 0, 0, 8]);
        Debug.Log(output_nn[0, 0, 0, 9]);
        worker.Dispose();
        return output_nn;
    }



    private Texture2D LoadByIO(string loadpath)
    {
        //FileStream creat
        FileStream fileStream = new FileStream(loadpath, FileMode.Open, FileAccess.Read);
        fileStream.Seek(0, SeekOrigin.Begin);
        //
        byte[] bytes = new byte[fileStream.Length];
        //FileStream read
        fileStream.Read(bytes, 0, (int)fileStream.Length);
        //FileStream close
        fileStream.Close();
        fileStream.Dispose();
        fileStream = null;

        // Ceate Texture
        int width = 28;
        int height = 28;
        Texture2D texture2D = new Texture2D(width, height);
        var is_load = texture2D.LoadImage(bytes);
        if (!is_load)
        {
            Debug.Log("Texture Fail");
        }
        return texture2D;
    }
    // Update is called once per frame
    void Update()
    {
        
    }
}
`

**Comment by KozykinRoman** [comment]
Are you going to recognize the number 5 or letter "S"?
I can see there are values of probabilities at the end in your screenshot.
ex., 0.09857323
You may have labels.txt file in the same directory with your pb file.
So you can notice which model or number has the best probability among those things.

**Comment by JetttZHENG** [comment]
> Are you going to recognize the number 5 or letter "S"?
> I can see there are values of probabilities at the end in your screenshot.
> ex., 0.09857323
> You may have labels.txt file in the same directory with your pb file.
> So you can notice which model or number has the best probability among those things.

Thanks for your reply.
It's the number 5.
The outputs from h5 model are  [1.8387052e-08, 7.7058097e-08, 2.8443156e-07, 1.5130955e-02,
        2.0308136e-08, 9.8478162e-01, 8.1684783e-07, 1.3152821e-06,
        5.4783020e-05, 3.0142777e-05].

**Comment by KozykinRoman** [comment]
Can we discuss via teamviewer now?

**Comment by JetttZHENG** [comment]
> Can we discuss via teamviewer now?

I have solved the my problem. Thank you!

It requires  the  input_layer name as the dictionary name for the inputs dictionary.

For my case, it should be 

`var inputs = new Dictionary<string, Tensor>();
inputs.Add("conv2d_1_input", tensor);`

And I found my in_put layer name from the verbose of **tensorflow_to_barracuda.py**.

**Comment by KozykinRoman** [comment]
that's pretty good. 
Then can you help me how to train keras model?
